### ISSUE TITLE
crc setup fails with setting up KVM on vanilla CentOS 7

### ISSUE THREAD

**Issue by cg2p** [code-ready/crc#1449]
### General information

  * OS: Linux 
  * Hypervisor: virtual private server (antsle) / LXC
  * Did you run `crc setup` before starting it (Yes/No)? No - attempting the setup process

## CRC version
```bash
CodeReady Containers version: 1.14.0+36ad776
OpenShift version: 4.5.4 (embedded in binary)
```
  
## CRC status
```bash
Machine 'crc' does not exist. Use 'crc start' to create it
```

## CRC config
```bash
```

## Host Operating System
```bash
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

### Steps to reproduce
1. as username who has sudo privileges
```
# vi /etc/sudoers
# add
<username>  ALL=(ALL) NOPASSWD:ALL
```
2. crc setup 
 

### Expected
Setup works

### Actual
crc setup fails
```
INFO Checking if oc binary is cached              
INFO Checking if podman remote binary is cached   
INFO Checking if goodhosts binary is cached       
INFO Checking if CRC bundle is cached in '$HOME/.crc' 
INFO Checking minimum RAM requirements            
INFO Checking if running as non-root              
INFO Checking if Virtualization is enabled        
INFO Checking if KVM is enabled                   
INFO Setting up KVM                               
FATA Failed to load kvm module  
```

**Comment by cfergeau** [comment]
Can you post the output of `crc setup --log-level debug` so that the actual failure can be seen? Can you make sure that virtualization is not disabled in your bios? `dmesg` will have a line about it if this is the reason for the failure.

**Comment by cg2p** [comment]
```
DEBU CodeReady Containers version: 1.14.0+36ad776 
DEBU OpenShift version: 4.5.4 (embedded in binary) 
INFO Checking if oc binary is cached              
DEBU Running '/home/<username>/.crc/bin/oc/oc version --client' 
DEBU oc binary already cached                     
INFO Checking if podman remote binary is cached   
DEBU Currently podman remote is not supported     
INFO Checking if goodhosts binary is cached       
DEBU goodhost binary already cached               
INFO Checking if CRC bundle is cached in '$HOME/.crc' 
INFO Checking minimum RAM requirements            
DEBU Total memory of system is 33458749440 bytes  
INFO Checking if running as non-root              
INFO Checking if Virtualization is enabled        
DEBU Checking if the vmx/svm flags are present in /proc/cpuinfo 
DEBU CPU virtualization flags are good            
INFO Checking if KVM is enabled                   
DEBU Checking if /dev/kvm exists                  
DEBU kvm kernel module is not loaded              
INFO Setting up KVM                               
DEBU Trying to load kvm module                    
DEBU exit status 1 :                              
FATA Failed to load kvm module
```

**Comment by cg2p** [comment]
`dmesg` gives a very large output - not sure what to check.

Do these outputs help?

```
$ lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                8
On-line CPU(s) list:   0-7
Thread(s) per core:    2
Core(s) per socket:    4
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 86
Model name:            Intel(R) Xeon(R) CPU D-1518 @ 2.20GHz
Stepping:              3
CPU MHz:               1750.708
BogoMIPS:              4400.39
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              6144K
NUMA node0 CPU(s):     0-7

$ egrep -wo 'vmx|ept|vpid|npt|tpr_shadow|flexpriority|vnmi|lm|aes' /proc/cpuinfo | sort | uniq
aes
ept
flexpriority
lm
tpr_shadow
vmx
vnmi
vpid

$ egrep -o '(vmx|svm)' /proc/cpuinfo | sort | uniq
vmx
```

**Comment by cg2p** [comment]
```
$ lsmod | grep -i kvm
kvm_intel             188644  0 
kvm                   621480  1 kvm_intel
irqbypass              13503  1 kvm
```

**Comment by praveenkumar** [comment]
```
DEBU Trying to load kvm module                    
DEBU exit status 1 :                              
```
Under the hood we run `modprobe kvm` command, if you run the `crc setup --log-level debug` you can find some more details around it.

**Comment by cg2p** [comment]
that is the `crc setup --log-level debug` output ^^^ from suggestion by @cfergeau - do you mean something different?

**Comment by praveenkumar** [comment]
@cg2p what is the output of following command on your system?

```
$ modprobe kvm
$ echo $?
```

**Comment by cg2p** [comment]
```
$ modprobe kvm
$ echo $?
1
```

**Comment by rseward** [comment]
I ran into a similar issue running CRC on an AMD Threadripper CPU on Centos 8. CRC wouldn't start for me. I solved my problem by dusting off an Intel Server with sufficient memory. CRC started with no problem on the Intel CPU with Centos 8. I spent about four to five days banging my head against the "apparent" problem with the Threadripper CPU. I don't know if your problem is the same, but if you can try an alternative environment it might improve your experience.

**Comment by cfergeau** [comment]
This will be addressed by https://github.com/code-ready/crc/pull/1454

**Comment by praveenkumar** [comment]
`kvm` module issue is fixed in latest release, closing this issue now.

**Comment by cg2p** [comment]
👍

**Comment by cg2p** [comment]
My issue is not resolved. Do I open a new issue?

I have a vanilla centOS 7.8 VM, running in LXC (my antsle private server). I don't know whether this is an issue / not possible for my environment.

e.g. CRC setup is trying to load KVM (all installed fine it seems) within the centOS 7.8 VM, which itself is in a LXC container on an edgeLinux machine - [antlets overview](https://docs.antsle.com/antlets/antlets)

Is this to do with some setting around **nested virtualization**?

The output of crc setup --log-level debug is:
```

DEBU CodeReady Containers version: 1.15.0+e317bed 
DEBU OpenShift version: 4.5.7 (embedded in binary) 
INFO Checking if oc binary is cached              
DEBU Running '/home/cg2p/.crc/bin/oc/oc version --client' 
DEBU oc binary already cached                     
INFO Checking if podman remote binary is cached   
DEBU Currently podman remote is not supported     
INFO Checking if goodhosts binary is cached       
DEBU goodhost binary already cached               
INFO Checking if CRC bundle is cached in '$HOME/.crc' 
INFO Checking minimum RAM requirements            
DEBU Total memory of system is 33458749440 bytes  
INFO Checking if running as non-root              
INFO Checking if Virtualization is enabled        
DEBU Checking if the vmx/svm flags are present in /proc/cpuinfo 
DEBU CPU virtualization flags are good            
INFO Checking if KVM is enabled                   
DEBU Checking if /dev/kvm exists                  
DEBU kvm kernel module is not loaded              
INFO Setting up KVM                               
DEBU Trying to load kvm module                    
INFO Will use root access: Load kvm_intel kernel module 
DEBU Running '/usr/bin/sudo modprobe kvm_intel'   
DEBU Command failed: exit status 1                
DEBU stdout:                                      
DEBU stderr:                                      
FATA Failed to load kvm intel module:  exit status 1:
```
and
```
$ modprobe kvm
$ echo $?
1
```
and
```
$ lsmod | grep -i kvm
kvm_intel             188644  0 
kvm                   621480  1 kvm_intel
irqbypass              13503  1 kvm
```

**Comment by cfergeau** [comment]
Does `/dev/kvm` exist? I'm assuming you did run `lsmod |grep -i kvm` within the centos 7.8 VM? Have you checked dmesg for potential hints?

**Comment by cg2p** [comment]
Yes `lsmod |grep -i kvm` was inside the guest VM.

Also, in guest VM:
```
$ lscpu | grep Virtualization
Virtualization:        VT-x
```
On the host bare metal edgeLinux machine (which is Antsle's custom centOS 7.8) KVM is install and `/dev/kvm/` exists.

But in the guest centOS VM `/dev/kvm` does not exist - even though I believe I have followed the kvm install into that guest VM:
```
$ yum install qemu-kvm libvirt libvirt-python libguestfs-tools virt-install
...ok
$ dmesg -c
$ systemctl enable libvirtd
Authorization not available. Check if polkit service is running or see debug message for more information.
$dmesg
<blank -- no output>
```

**Comment by cfergeau** [comment]
> But in the guest centOS VM /dev/kvm does not exist - even though I believe I have followed the kvm install into that guest VM:

This is what is causing crc to fail then. This should be created when the kvm kernel modules are loaded. You could unload `kvm_intel`, reload it, and check `dmesg`

**Comment by cg2p** [comment]
SOLVED

I had not edited the config of my guest VM to allow pass through to the underlying host hypervisor. Fixed it through the virsh utility following [this](https://ostechnix.com/how-to-enable-nested-virtualization-in-kvm-in-linux/ ).

Once in place `virt-host-validate` shows a PASS for does `/dev/kvm/` exist check. 

Many thanks for help.
